hardware_interface/include/hardware_interface/handle.hpp

```diff
@@ -16,19 +16,126 @@
 #define HARDWARE_INTERFACE__HANDLE_HPP_
 
 #include <string>
+#include <vector>
 #include <utility>
 
 #include "hardware_interface/macros.hpp"
 #include "hardware_interface/visibility_control.h"
 
+#include <hardware_interface/hardware_info.hpp>
+
 namespace hardware_interface
 {
+struct InterfaceConfiguration
+{
+  std::string joint_name;
+  InterfaceInfo interface_info;
+};
+
+/// A variant class used to store value on a given interface
+class Variant
+{
+enum value_type
+{
+  boolean = 0,
+  integer,
+  floating,
+  vector_boolean,
+  vector_integer,
+  vector_floating,
+};
+
+public:
+  Variant(bool value)
+  {
+    value_type_ = value_type::boolean;
+    bool_value_ = value;
+  }
+
+  Variant(int value)
+  {
+    value_type_ = value_type::integer;
+    int_value_ = value;
+  }
+
+  Variant(double value)
+  {
+    value_type_ = value_type::floating;
+    double_value_ = value;
+  }
+
+  Variant(const std::vector<bool> & value)
+  {
+    value_type_ = value_type::vector_boolean;
+    vector_bool_value_ = value;
+  }
+
+  Variant(const std::vector<int> & value)
+  {
+    value_type_ = value_type::vector_integer;
+    vector_int_value_ = value;
+  }
+
+  Variant(const std::vector<double> & value)
+  {
+    value_type_ = value_type::vector_floating;
+    vector_double_value_ = value;
+  }
+
+  void set_value(bool value)
+  {
+    bool_value_ = value;
+  }
+
+  void set_value(int value)
+  {
+    int_value_ = value;
+  }
+
+  void set_value(double value)
+  {
+    double_value_ = value;
+  }
+
+  template <typename DataT, typename std::enable_if<std::is_integral<DataT>, bool> = true>
+  DataT get_value()
+  {
+    return bool_value_;
+  }
+
+  template <typename DataT, typename std::enable_if<std::is_integral<DataT>, int> = true>
+  DataT get_value()
+  {
+    return int_value_;
+  }
+
+  template <typename DataT, typename std::enable_if<std::is_floating_point<DataT>, double> = true>
+  DataT get_value()
+  {
+    return double_value_;
+  }
+
+  bool is_valid()
+  {
+    // Check if nan or empty to return "false"
+  }
+
+private:
+  value_type value_type_;
+  bool bool_value_;
+  int int_value_;
+  double double_value_;
+  std::vector<bool> vector_bool_value_;
+  std::vector<int> vector_int_value_;
+  std::vector<double> vector_double_value_;
```

Comment:
Maybe you should add a member that would store whether this interface is initialized and a getter function.

`bool is_initialized_;`
`
bool is_initialized()
{
  return is_initialized_;
}
`

Reply:
I would really like to be rid of initialized flag values and instead use RAII like it was meant to be used. All constructed objects are either valid or don't exist. We should use the type system for things like this.